presto-cli/src/main/java/com/facebook/presto/cli/StatusPrinter.java

```diff
@@ -382,8 +392,19 @@ private void printStageTree(StageStats stage, String indent, AtomicInteger stage
             rowsPerSecond = formatCountRate(0, new Duration(0, SECONDS), false);
         }
         else {
-            bytesPerSecond = formatDataRate(bytes(stage.getProcessedBytes()), elapsedTime, false);
-            rowsPerSecond = formatCountRate(stage.getProcessedRows(), elapsedTime, false);
+            ProgressStatus previousStage;
+            if (previousStageStatuses.containsKey(id)) {
+                previousStage = previousStageStatuses.get(id);
+            }
+            else {
+                previousStage = new ProgressStatus(start, 0, 0);
+            }
+            Duration elapsedTime = succinctNanos(now - previousStage.getNanoTime());
+
+            bytesPerSecond = formatDataRate(bytes(stage.getProcessedBytes() - previousStage.getProcessedBytes()), elapsedTime, false);
+            rowsPerSecond = formatCountRate(stage.getProcessedRows() - previousStage.getProcessedRows(), elapsedTime, false);
+
+            previousStageStatuses.put(id, new ProgressStatus(now, stage.getProcessedBytes(), stage.getProcessedRows()));
```

Comment:
Instead of overwriting the same map, I personally prefer to have a new `currentStageStatuses` and passing it into the method, and do the switch:
```
previousStageStatuses = currentStageStatuses;
```
when everything is finished.

Another thought would be to use `ImmutableMap` and `ImmutableMap.Builder`.

Will let @raghavsethi  decide :)